### PR TITLE
Feature/websocket

### PR DESCRIPTION
**Pull request by xrdavies**
* implement websocket

Use wslay which is MIT licensed as websocket frame/event handler.
When there is new block applied, the pool will broadcast this block info to all connected websocket clients. Currently, the websocket service is started with RPC service.

### CODE REVIEW

File: client/websocket/websocket.c [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+			LL_APPEND(g_session_list, elem);
+			pthread_mutex_unlock(&g_mutex_sessions);
+			ws_info("%s connected\n", inet_ntoa(session->addr));
+		}
+	}
+}
+
+static void sha1(uint8_t *dst, const uint8_t *src, size_t src_length)
+{
+	SHA_CTX ctx;
```
Comment (jonano614) [inline client/websocket/websocket.c]: Do we really need SHA-1? This algorithm is a bit outdated.

File: client/block.c [hunk 2]
```diff
@@ -1540,6 +1543,61 @@ const char* xdag_get_block_state_info(uint8_t flags)
...
+			, bi->flags & ~BI_OURS
+			, xdag_get_block_state_info(bi->flags)
+			, (unsigned long long)h[3], (unsigned long long)h[2], (unsigned long long)h[1], (unsigned long long)h[0]
+			, xdag_diff_args(bi->difficulty)
+			, get_remark(bi)
+			, address
+			, pramount(bi->amount)
+			);
+
+	if((bi->flags & BI_REF) && bi->ref != NULL) {
```
Comment (sgaragagghu) [inline client/block.c]: there can be a segfault here

File: client/utils/uuid.c [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+static uint64_t xorshift128plus(uint64_t *s) {
+	uint64_t s1 = s[0];
+	uint64_t s0 = s[1];
+	s[0] = s0;
+	s1 ^= s1 << 23;
+	s[1] = s1 ^ s0 ^ (s1 >> 18) ^ (s0 >> 5);
+	return s[1] + s0;
+}
+
+int uuid(char *buf)
```
Comment (sgaragagghu) [inline client/utils/uuid.c]: cannot reuse our PRG?

File: client/websocket/websocket.c (hunk 1 above)
Comment (sgaragagghu) [reply client/websocket/websocket.c]: websocket protocol uses sha-1

File: client/utils/base64.c [hunk 4]
```diff
@@ -0,0 +1,122 @@
+//
```
[the remark this answers is not in the document]
Comment (xrdavies) [reply client/utils/base64.c]: It's different. The implementation in address.c is just for 32 bytes length, can not be used commonly as far as I know, no padding.

File: client/websocket/websocket.c (hunk 1 above)
Comment (xrdavies) [reply client/websocket/websocket.c]: It's needed as described in RFC6455. Just old one.

File: client/utils/uuid.c (hunk 3 above)
Comment (xrdavies) [reply client/utils/uuid.c]: Yes, good point. Meant to reuse our RPG, but when I implemented this code, we still donot have RPG yet.

File: client/block.c (hunk 2 above)
Comment (xrdavies) [reply client/block.c]: Do you mean the old issue of the extra blocks? I think this function only called in block.c in the same thread as add_block_nolock.

File: client/utils/uuid.c (hunk 3 above)
Comment (xrdavies) [reply client/utils/uuid.c]: Will try to replace it later.

File: client/block.c (hunk 2 above)
Comment (sgaragagghu) [reply client/block.c]: there are many threads in block.c that can change those values, and this function isn't executed by some thread about websocket?

Comment (xrdavies) [reply client/block.c]: Not in some thread about web socket, but still in the same thread of add_block_nolock. web socket thread will fetch message from a queue which is protected by mutex.

Comment (xrdavies) [reply client/block.c]: @sgaragagghu  Updated.

**Comment by xrdavies** [comment]
@jonano614 @sgaragagghu  Any more comments?

**Pull request merged**